### PR TITLE
MySQL Create Connection Stub Implementation

### PR DESCRIPTION
**Pull request by rjobanp**
### Motivation

Implement `CREATE CONNECTION .. TO MYSQL..` to allow continued prototype development. TLS / PrivateLink / SSH Tunnel options are defined but not currently implemented. This functionality sits behind a feature-flag.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/coord/ddl.rs
Comment (petrosagg) [inline src/adapter/src/coord/ddl.rs:1070]: I'm a bit nervous that we'll forget something if it only exists in TODO comments. I suggest we both leave a TODO in the code as you have here and also adding whatever it is to be done in the work item list in this issue https://github.com/MaterializeInc/materialize/issues/15901

I added these ones already

File: src/sql/src/plan/statement/ddl.rs
Comment (petrosagg) [inline src/sql/src/plan/statement/ddl.rs:3407]: let's use a general named flag for the whole MySQL feature that everything will be gated with

File: src/storage-types/src/connections.proto
Comment (petrosagg) [inline src/storage-types/src/connections.proto:133]:
```suggestion
    ProtoTunnel tunnel = 9;
```

File: src/storage-types/src/connections.rs
Comment (petrosagg) [inline src/storage-types/src/connections.rs:1440]: Since this is a separate type it would be better to provide a standalone `impl RustType<ProtoMySqlSslMode> for MySqlSslMode` implementation and use `proto.tls_mode.into_rust_if_some("...")?,` here

**Review by petrosagg — COMMENTED** [review]
Very nice!

**Review by petrosagg — APPROVED** [review]

**Review by jkosh44 — APPROVED** [review]
Adapter code LGTM

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (benesch) [inline src/sql/src/plan/statement/ddl/connection.rs:378]: MySQL seems to like these in all UPPERCASE. I recommend that we call `.to_uppercase()` here so that we can use the MySQL preferred spelling in code (`DISABLED`, `REQUIRED`, `VERIFY_CA`, `VERIFY_IDENTITY`)—and also in our docs—but users can case things however they want and it'll work.

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (benesch) [inline src/sql/src/plan/statement/ddl/connection.rs:384]: I like that you're accepting kebab and snake case here!

File: src/testdrive/src/action/mysql/connect.rs
Comment (rjobanp) [reply src/testdrive/src/action/mysql/connect.rs:43]: Agreed, I spent all day yesterday struggling to replicate the error on this -- the [healthcheck](https://github.com/MaterializeInc/materialize/blob/main/misc/python/materialize/mzcompose/services/mysql.py#L41) appears to work fine and is the [recommended](https://stackoverflow.com/questions/42567475/docker-compose-check-if-mysql-connection-is-ready) way to validate mysql is ready, but for some reason the first testdrive `mysql-connect` call would hit a 'Connection Refused' error on the first attempt when run on buildkite but not locally.

Then I noticed that the `postgres-connect` testdrive client [implemented](https://github.com/MaterializeInc/materialize/blob/main/src/testdrive/src/util/postgres.rs#L56) this Retry behavior so I copied over the same implementation to make this work.

However I've now since removed that `mysql-connect` call from the test-case since we removed the `database` option. I could remove this retry behavior but I suspect that the next test-case we add will hit the same thing. I have a feeling something is different about the docker environment in buildkite vs my local one

Comment (rjobanp) [reply src/testdrive/src/action/mysql/connect.rs:43]: Actually I just realized I'm hitting this again now that I removed the `mysql-connect` call and the first connection to MySQL is the Create Connection validation https://buildkite.com/materialize/tests/builds/71613#018c8435-29e8-456f-8d67-a360c620d38f which doesn't have this retry 🤦🏽 . Will keep digging

File: src/storage-types/src/connections.rs
Comment (rjobanp) [reply src/storage-types/src/connections.rs:1275]: I was matching the MySQL spelling for these options which uses `DISABLED` `REQUIRED`, `VERIFY_CA`, and `VERIFY_IDENTITY`, etc

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (rjobanp) [reply src/sql/src/plan/statement/ddl/connection.rs:384]: stolen from the postgres connection implementation!

**Review by def- — COMMENTED** [review]
I assume this is currently not well testable?

**Comment by rjobanp** [comment]
> I assume this is currently not well testable?

@def- I added a simple testdrive test, but most of the connection options are not fully implemented (e.g. SSL, SSH Tunneling, Privatelink). We're aiming to build an e2e prototype of the MySQL source asap so the plan is to fill out those pieces once we've finished the prototype. I assume that's what you're referring to?

File: src/adapter/src/catalog/state.rs
Comment (def-) [inline src/adapter/src/catalog/state.rs:2221]: Uncovered

File: src/sql/src/plan/statement/ddl.rs
Comment (def-) [inline src/sql/src/plan/statement/ddl.rs:4873]: Uncovered

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (def-) [inline src/sql/src/plan/statement/ddl/connection.rs:374]: Uncovered

File: src/sql/src/plan/statement/ddl/connection.rs
Comment (def-) [inline src/sql/src/plan/statement/ddl/connection.rs:392]: Uncovered

File: src/testdrive/src/action/mysql/connect.rs
Comment (def-) [inline src/testdrive/src/action/mysql/connect.rs:39]: Uncovered

**Review by def- — COMMENTED** [review]
Yes, thank you. I ran coverage anyway, maybe we can refer to it later when adding testing: https://buildkite.com/materialize/coverage/builds/328#018c87fc-57f1-44dd-8a6a-0de50980967b (inline in comments)

File: src/testdrive/src/action/mysql/connect.rs
Comment (rjobanp) [reply src/testdrive/src/action/mysql/connect.rs:41]: Figured it out! I updated the healthcheck to use a TCP connection for validation -- it was defaulting to using a unix-socket to check if MySQL was ready to accept connections which must happen quicker than when using TCP.

Comment (rjobanp) [reply src/testdrive/src/action/mysql/connect.rs:41]: I think it may still make sense to leave this retry logic here to match the postgres behavior, wdyt?

**Review by morsapaes — APPROVED** [review]

File: src/testdrive/src/action/mysql/connect.rs
Comment (benesch) [reply src/testdrive/src/action/mysql/connect.rs:41]: ~I think the healthcheck is broken! According to this SO post `mysqladmin --ping` can return a false positive if the server is running but not accepting connections: https://stackoverflow.com/questions/42567475/docker-compose-check-if-mysql-connection-is-ready#comment75589721_42757250~

~The timestamps seem to bear that out. In services.log...~

```
mysql-cdc-mysql-1         | 2023-12-19T22:52:53.631782Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Bind-address: '::' port: 33060, socket: /var/run/mysqld/mysqlx.sock
mysql-cdc-mysql-1         | 2023-12-19T22:52:53.631857Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.35'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server - GPL.
```

~In testdrive...~

```
> CREATE SECRET [query truncated on purpose so as to not reveal the secret in the log]
rows match; continuing at ts 1703026372.2337284
> CREATE CONNECTION mysq TO MYSQL (
  HOST mysql,
  USER root,
  PASSWORD SECRET mysqlpass
)
query error matches; continuing
Connecting to PostgreSQL server at postgres://mz_system:materialize@materialized:6877...
>> ALTER SYSTEM SET enable_mysql_source = true
> CREATE CONNECTION mysq TO MYSQL (
  HOST mysql,
  USER root,
  PASSWORD SECRET mysqlpass
)
```

~1703026372.2337284 is 22:52:52.233, so testdrive is trying to connect (or, at least, is a few commands away from trying to connect) 1.5 seconds before MySQL is accepting connections.~

Scratch all the above; I wrote it up before you mentioned you'd tracked it down. `--protocol=TCP` is a great, great catch—nice work! 🙇🏽 

> I think it may still make sense to leave this retry logic here to match the postgres behavior, wdyt?

I'd leave it without the retry. @def- recently added the retry loop in https://github.com/MaterializeInc/materialize/pull/22300/files#r1354183675 and I'm not entirely sure why! @def-, is it possible that you could use `tokio::time::timeout` in the PostgreSQL connection logic rather than `retry_async_canceling`?

File: src/storage-types/src/connections.rs
Comment (benesch) [reply src/storage-types/src/connections.rs:1275]: Ah, fair enough. These are internal only but I suppose it's worth aligning them with the public API.

**Comment by benesch** [comment]
Don't wait for my approval to merge! I defer to those who have looked more closely. (LGTM from a SQL council perspective though.)

**Pull request merged**
